src/tools/illink/src/linker/Linker/RuntimeInterfaceImplementation.cs

```diff
@@ -0,0 +1,101 @@
+	{
+		/// <summary>
+		/// The type that implements <see cref="RuntimeInterfaceImplementation.InflatedInterfaceType"/>.
+		/// </summary>
+		public TypeDefinition Implementor { get; }
+
+		/// <summary>
+		/// All the chains of .interfaceImpls that cause <see cref="Implementor"/> to implement <see cref="InflatedInterfaceType"/>
+		/// </summary>
+		public ImmutableArray<InterfaceImplementationChain> InterfaceImplementationChains { get; }
+
+		/// <summary>
+		/// The inflated interface type reference that is implemented by <see cref="Implementor"/>.
+		/// </summary>
+		public TypeReference InflatedInterfaceType { get; }
+
+		/// <summary>
+		/// The <see cref="TypeDefinition"/> of <see cref="InflatedInterfaceType"/>
+		/// </summary>
+		public TypeDefinition? InterfaceTypeDefinition { get; }
+
+		public RuntimeInterfaceImplementation (TypeDefinition implementor, TypeReference interfaceType, TypeDefinition? interfaceTypeDefinition, IEnumerable<InterfaceImplementationChain> interfaceImplementations)
+		{
+			Implementor = implementor;
+			InterfaceImplementationChains = interfaceImplementations.ToImmutableArray ();
+			InflatedInterfaceType = interfaceType;
+			InterfaceTypeDefinition = interfaceTypeDefinition;
+		}
+
+		public bool IsAnyImplementationMarked (AnnotationStore annotations, ITryResolveMetadata context)
+		{
+			if (annotations.IsMarked (this))
+				return true;
+			foreach (var chain in InterfaceImplementationChains) {
+				if (chain.IsMarked (annotations, context)) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		/// <summary>
+		/// Returns a list of InterfaceImplementationChains for a derived type of <see cref="Implementor"/>.
+		/// </summary>
+		public IEnumerable<(TypeReference InterfaceType, InterfaceImplementationChain Chain)> CreateImplementationChainsForDerivedType (TypeReference baseTypeRef, ITryResolveMetadata context)
+		{
+			// This is only valid for classes
+			Debug.Assert (Implementor.IsClass);
+			Debug.Assert (Implementor == context.TryResolve (baseTypeRef));
+
+			var inflatedInterfaceType = InflatedInterfaceType.TryInflateFrom (baseTypeRef, context);
+			Debug.Assert (inflatedInterfaceType is not null);
+
+			foreach (var impl in InterfaceImplementationChains) {
+				var inflatedImplProvider = impl.TypeWithInterfaceImplementation.TryInflateFrom (baseTypeRef, context);
+				Debug.Assert (inflatedImplProvider is not null);
+
+				yield return (inflatedInterfaceType, new InterfaceImplementationChain (inflatedImplProvider, impl.InterfaceImplementations));
+			}
+		}
+
+		/// <summary>
+		/// Returns a list of InterfaceImplementationChains for a type that has an explicit implementation of <see cref="Implementor"/>.
+		/// </summary>
+		public IEnumerable<(TypeReference InterfaceType, InterfaceImplementationChain Chain)> CreateImplementationChainForImplementingType (TypeDefinition typeThatImplementsImplementor, InterfaceImplementation impl, ITryResolveMetadata context)
```

Comment:
I think I meant this to be `typeThatImplementImplementor`. This parameter should be a type that has an explicit implementation the `Implementor` property. So if IFoo implements IBase and C implements IFoo, then this would be called on the `RuntimeInterfaceImplementation` that represents `<IFoo implements IBase>`, and the parameters should be (the `TypeDefinition` for `C`) and (the InterfaceImplementation representing C's .interfaceImpl pointing to `IFoo`). I don't love the name and am happy to change it, but I don't think `typeThatImplementsInterface` is quite right either.

Reply:
I ended up moving the runtime interfaces algorithm / cache into its own type and inlined these helpers, so there shouldn't be a naming issue here anymore.